server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -256,44 +256,52 @@ public boolean equalsAliases(MetaData other) {
      * present for that index
      */
-    public ImmutableOpenMap<String, List<AliasMetaData>> findAliases(final String[] aliases, String[] concreteIndices) {
+    public ImmutableOpenMap<String, List<AliasMetaData>> findAliases(String[] aliases, String[] concreteIndices) {
+        return findAliases(false, aliases, concreteIndices);
+    }
+
+    /**
+     * Finds the specific index aliases that match with the specified aliases directly or partially via wildcards and
+     * that point to the specified concrete indices or match partially with the indices via wildcards.
+     *
+     * @param aliasesProvided Whether concrete aliases where provided by the user in the original request. This information
+     *                        gets lost as internally aliases may get replaced.
+     * @param aliases         The names of the index aliases to find
+     * @param concreteIndices The concrete indexes the index aliases must point to order to be returned.
+     * @return a map of index to a list of alias metadata, the list corresponding to a concrete index will be empty if no aliases are
+     * present for that index
+     */
+    public ImmutableOpenMap<String, List<AliasMetaData>> findAliases(boolean aliasesProvided, String[] aliases, String[] concreteIndices) {
         assert aliases != null;
         assert concreteIndices != null;
         if (concreteIndices.length == 0) {
             return ImmutableOpenMap.of();
         }
 
-        boolean matchAllAliases = matchAllAliases(aliases);
         ImmutableOpenMap.Builder<String, List<AliasMetaData>> mapBuilder = ImmutableOpenMap.builder();
         Iterable<String> intersection = HppcMaps.intersection(ObjectHashSet.from(concreteIndices), indices.keys());
         for (String index : intersection) {
             IndexMetaData indexMetaData = indices.get(index);
             List<AliasMetaData> filteredValues = new ArrayList<>();
             for (ObjectCursor<AliasMetaData> cursor : indexMetaData.getAliases().values()) {
                 AliasMetaData value = cursor.value;
-                if (matchAllAliases || Regex.simpleMatch(aliases, value.alias())) {
+                if (Strings.isAllOrWildcard(aliases) || Regex.simpleMatch(aliases, value.alias())) {
                     filteredValues.add(value);
                 }
             }
 
             if (!filteredValues.isEmpty()) {
                 // Make the list order deterministic
                 CollectionUtil.timSort(filteredValues, Comparator.comparing(AliasMetaData::alias));
+                mapBuilder.put(index, Collections.unmodifiableList(filteredValues));
+            } else if (aliasesProvided == false) {
+                // in case all aliases are requested then it is desired to return the concrete index with no aliases (#25114):
+                mapBuilder.put(index, Collections.emptyList());
```

Comment:
given how weird this API is, I wonder if we could keep this logic contained in TransportGetAliasesAction, similar to what we have in RestGetAliasesAction, rather than introducing this flag in MetaData and exposing it to potentially other API.

Reply:
Makes sense. I will try to move it to TransportGetAliasesAction.